source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/UtilityFunctions/SeriesDifferenceUtilityFunction.py

```diff
@@ -0,0 +1,98 @@
+from mlos.Spaces import SimpleHypergrid, ContinuousDimension, CategoricalDimension, Point
+from mlos.Spaces.Configs.ComponentConfigStore import ComponentConfigStore
+from mlos.Tracer import trace
+
+
+series_difference_utility_function_config_store = ComponentConfigStore(
+    parameter_space=SimpleHypergrid(
+        name="series_difference_utility_function_config",
+        dimensions=[
+            # I might use some of these from ConfidenceBoundUtilityFunction... They seem pretty reasonable.
+            CategoricalDimension(name="utility_function_name", values=["lower_confidence_bound_on_improvement", "upper_confidence_bound_on_improvement"]),
+            ContinuousDimension(name="alpha", min=0.01, max=0.5)
+        ]
+    ),
+    default=Point(
+        utility_function_name="upper_confidence_bound_on_improvement",
+        alpha=0.01
+    )
+)
+
+
+class SeriesDifferenceUtilityFunction(UtilityFunction):
+    def __init__(self, function_config: Point, surrogate_model: MultiObjectiveRegressionModel, objective: SeriesObjective, logger=None):
+        if logger is None:
+            logger = create_logger(self.__class__.__name__)
+        self.logger = logger
+
+        self.config = function_config
+        self.series_objective = objective
+        self._sign = 1 if not self.series_objective.minimize else -1
+        self.surrogate_model: MultiObjectiveRegressionModel = surrogate_model
+
+    @trace()
+    def __call__(self, feature_values_pandas_frame):
+        # NOTE: THIS LOGIC IS DUPLICATED INTO SeriesPredictedValueUtilityFunction and SeriesDifferenceUtilityFunction.
+        # Later they will diverge as SeriesDifferenceUtilityFunction should require additional logic surrounding probability
+        # of improvement. Perhaps this part should be refactored to be shared though.
+        #
+        self.logger.debug(f"Computing utility values for {len(feature_values_pandas_frame.index)} points.")
+
+        predicted_value_col = Prediction.LegalColumnNames.PREDICTED_VALUE.value
+        predicted_value_var_col = Prediction.LegalColumnNames.PREDICTED_VALUE_VARIANCE.value
+        dof_col = Prediction.LegalColumnNames.PREDICTED_VALUE_DEGREES_OF_FREEDOM.value
+
+        del feature_values_pandas_frame[f"context_space.{self.series_objective.series_modulation_dimension.name}"]
+        series_vals_df = pd.DataFrame({
+            f"context_space.{self.series_objective.series_modulation_dimension.name}": self.series_objective.series_modulation_dimension.linspace()
+        })
+        feature_values_pandas_frame_merged = feature_values_pandas_frame.merge(series_vals_df, how="cross")
+        multi_objective_predictions = self.surrogate_model.predict(features_df=feature_values_pandas_frame_merged)
+
+        predictions = multi_objective_predictions[0]
+        predictions_df = predictions.get_dataframe()
+
+        # TODO ZACK: Vectorize this. I'm sure pandas/numpy can do this faster. @Adam, any suggestions
+        resulting_series_arry = []

```

Comment:
array?

Reply:
What is resulting_series_arry? What does it hold? Also, why is series_modulation_dimension categorical?

Reply:
Got it, it's a predicted value series

Reply:
series_modulation_dimension was my best idea for how to name the dimension which is modulated across the time series. I chose categorical so that I could feed it a set of values that are of interest. For example, if kusto gives performance metrics at t=0.1, t=0.5, and t=1s, it is foolish to check any other value and try and compare the resulting series. The only dimension type that I thought of that would work in this case is categorical. Predicted_value_series seems like that is the result of the prediction when it is the opposite, it is the input to the prediction. 

Do you have any thoughts for the name now? I thought about just using something like time_dimension but as I will elaborate on later, I think that this functionality could be used for more complex things than just time-modulation and the time-convention doesn't really add that much readability in my mind. It is, to me, just the ability to test across a wide sample of values and converge on a specified series.